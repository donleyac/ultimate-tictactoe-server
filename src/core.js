import {Iterable, fromJS, List, Map,is} from 'immutable';

export const INITIAL_STATE = getInitial();
export function getInitial() {
  let board = [...Array(3).keys()].map(i => Array(3).fill(0));
  board.forEach(row=>{
    row.forEach((elem,index)=>{
      row[index] = {
        board: fromJS([...Array(3).keys()].map(i => Array(3).fill(0))),
        selectable: true,
        winner: 0
      }
    })
  })
  return Map({
    board: fromJS(board),
    winner: 0,
    player: 1
  });
}
export function placePiece(state, grid, cell, playerId) {
  if(playerId===state.get("player") && state.get("board").get(grid[0]).get(grid[1]).get("board").get(cell[0]).get(cell[1])===0){
    //Place Piece in cells
    let placed = state.updateIn(
      ["board", grid[0], grid[1], "board", cell[0], cell[1]],
      0,
      board => playerId
    );
    //Set selectable
    let selectable = placed.updateIn(
      ["board"],
      0,
      board => {
        let new_board = [...Array(3).keys()].map(i => Array(3).fill(0));
        board.forEach((row,r_index)=>{
          row.forEach((elem,index)=>{
            let grid = board.get(r_index).get(index);
            if(r_index!==cell[0] || index!==cell[1] && board.get(cell[0]).get(cell[1]).get("winner")===0){
              new_board[r_index][index]={board:grid.get("board"), selectable: false, winner: grid.get("winner")};
            }
            else {
              new_board[r_index][index]={board:grid.get("board"), selectable: true, winner: grid.get("winner")};
            }
          });
        });
        return fromJS(new_board);
      }
    );
    //Check Winner in Grid
    let grid_winner = selectable.updateIn(
      ["board", grid[0], grid[1]],
      0,
      grid => {
        let board = grid.get("board");
        let winner = checkWinner(board);
        return Map({board: board, selectable: grid.get("selectable"), winner: winner});
      }
    );
    //Check Game Winner and change player
    let game_winner = grid_winner.updateIn(
      [],
      0,
      state => {
        let board = state.get("board");
        let winner = checkWinner(board);
        return Map({board: board, winner: winner, player:playerId*-1});
      }
    );
    return game_winner;
  }
  return {response: "Invalid Click"};
}
function checkWinner(board) {
  if(Object(board.get(0).get(0))===board.get(0).get(0)) {    //Horizontal Check
    for(let i=0; i<board.size; i++){
      let row = board.get(i);
      if(Math.abs(row.get(0).get("winner")+row.get(1).get("winner")+row.get(2).get("winner"))===3){
        return row.get(0).get("winner");
      }
    }
    //Vertical Check
    for(let i=0; i<board.size; i++){
      if(Math.abs(board.get(0).get(i).get("winner")+board.get(1).get(i).get("winner")+board.get(2).get(i).get("winner"))===3){
        return board.get(0).get(i).get("winner");
      }
    }
    //Diagonal Check
    if(Math.abs(board.get(0).get(0).get("winner")+board.get(1).get(1).get("winner")+board.get(2).get(2).get("winner"))===3){
      return board.get(0).get(0).get("winner");
    }
    if(Math.abs(board.get(0).get(2).get("winner")+board.get(1).get(1).get("winner")+board.get(2).get(0).get("winner"))===3){
      return board.get(0).get(2).get("winner");
    }
  }
  else {
    //Horizontal Check
    for(let i=0; i<board.size; i++){
      let row = board.get(i);
      if(Math.abs(row.get(0)+row.get(1)+row.get(2))===3){
        return row.get(0);
      }
    }
    //Vertical Check
    for(let i=0; i<board.size; i++){
      if(Math.abs(board.get(0).get(i)+board.get(1).get(i)+board.get(2).get(i))===3){
        return board.get(0).get(i);
      }
    }
    //Diagonal Check
    if(Math.abs(board.get(0).get(0)+board.get(1).get(1)+board.get(2).get(2))===3){
      return board.get(0).get(0);
    }
    if(Math.abs(board.get(0).get(2)+board.get(1).get(1)+board.get(2).get(0))===3){
      return board.get(0).get(2);
    }
  }
  return 0;
}

// export function modIndicator(state, playerId, label, value, op){
//   if (op==="replace"){
//     return state.updateIn(
//       ["playersById",playerId, 'indicators', label],
//       0,
//       indicator => value
//     );
//   }
//   else {
//     return state.updateIn(
//       ["playersById",playerId, 'indicators', label],
//       0,
//       indicator => indicator + value
//     );
//   }
// }
// export function modCollection(state, collect, prop, val, op){
//   //need to cast val to immutable for add and comparisons
//   if(val instanceof Array){
//     val = List(val);
//   }
//   let findIndex = function(curr,target){
//     return curr.findIndex(function(elem){
//       if(Iterable.isIterable(elem)){
//         return is(elem.get(5), target.get(5));
//       }
//       else {
//         return is(elem, target);
//       }
//     })
//   }
//   let remove = function(curr, target){
//     return curr.delete(findIndex(curr,target));
//   }
//   //TODO Method is already generic, may be able to work with just a unique operation
//   if(state.getIn(["collections",collect,"layout"])==="free" && prop==="content" && op==="chg"){
//     return state.updateIn(
//       ["collections", collect, prop],
//       0,
//       content => content.set(findIndex(content, val), val));
//   }
//   else {
//     if(op==="add"){
//       return state.updateIn(
//        ["collections", collect, prop],
//        0,
//        content => content.push(val));
//     }
//     else if(op==="rm"){
//       return state.updateIn(
//         ["collections", collect, prop],
//         0,
//         content => remove(content, val));
//     }
//     //Replace list, mainly for re-ordering
//     else {
//       return state.updateIn(
//         ["collections", collect, prop],
//         0,
//         content => val);
//     }
//   }
// }
