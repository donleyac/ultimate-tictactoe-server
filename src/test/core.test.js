import {fromJS, List, Map} from 'immutable';
import config from './initial.json';
import joinGameST from './states/joinGame.json';
import placePieceST from './states/placePiece.json';
import {getInitial, joinGame, placePiece, switchPlayer, createRoom, joinRoom} from '../core.js';

const initialST = getInitial();
const createRoomST = createRoom(initialST, "apple", "alex");
const joinRoomST = joinRoom(createRoomST, "apple", "john");


test('Initial State', () => {
  expect(getInitial()).toMatchSnapshot();
});
test('Create Room', ()=>{
  //Standard create room
  expect(createRoomST).toMatchSnapshot();
  //Create Room that already existingUsers
  expect(createRoom(createRoomST, "apple", "john")).toMatchSnapshot();
});
test('Join Game', () => {
  expect(joinGame(fromJS(joinGameST), "apple", "john")).toMatchSnapshot();
});
test('Place Piece', ()=> {
  expect(placePiece(fromJS(placePieceST), "apple", 0, 0, 1)).toMatchSnapshot();
})
test('switchPlayer', ()=> {
  expect(switchPlayer(fromJS(placePieceST))).toMatchSnapshot();
})
