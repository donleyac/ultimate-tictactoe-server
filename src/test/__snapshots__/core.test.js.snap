exports[`test Create Room 1`] = `
Object {
  "rooms": Object {
    "apple": Object {
      "users": Array [
        "alex",
      ],
    },
  },
}
`;

exports[`test Create Room 2`] = `null`;

exports[`test Initial State 1`] = `
Object {
  "rooms": Object {},
}
`;

exports[`test Join Game 1`] = `
Object {
  "room": "apple",
  "rooms": Object {
    "apple": Object {
      "game": Object {
        "activePlayer": 1,
        "board": Array [
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 0,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 0,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 0,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 0,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 0,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 0,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 0,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 0,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 0,
          },
        ],
        "minPlayers": 2,
        "players": Object {
          "alex": -1,
          "john": 1,
        },
        "winner": 0,
      },
      "users": Array [
        "alex",
        "john",
      ],
    },
  },
  "username": "alex",
}
`;

exports[`test Place Piece 1`] = `
Object {
  "room": "apple",
  "rooms": Object {
    "apple": Object {
      "game": Object {
        "activePlayer": 1,
        "board": Array [
          Object {
            "grid": Array [
              1,
              1,
              1,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": false,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": false,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": false,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": false,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": false,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": false,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": false,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": false,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": false,
            "winner": 1,
          },
        ],
        "minPlayers": 2,
        "players": Object {
          "alex": -1,
          "john": 1,
        },
        "winner": 1,
      },
      "users": Array [
        "alex",
        "john",
      ],
    },
  },
  "username": "john",
}
`;

exports[`test switchPlayer 1`] = `
Object {
  "room": "apple",
  "rooms": Object {
    "apple": Object {
      "game": Object {
        "activePlayer": 1,
        "board": Array [
          Object {
            "grid": Array [
              0,
              1,
              1,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 0,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 1,
          },
          Object {
            "grid": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "selectable": true,
            "winner": 1,
          },
        ],
        "minPlayers": 2,
        "players": Object {
          "alex": -1,
          "john": 1,
        },
        "winner": 0,
      },
      "users": Array [
        "alex",
        "john",
      ],
    },
  },
  "username": "john",
}
`;
